common/kvstore/KeyValueStore.h

```diff
@@ -41,15 +38,47 @@ class KeyValueStore {
      * any entries, but each will see their own set of values.
      */
     KeyValueStore(std::string version, std::string path, std::string flavor);
+    ~KeyValueStore() noexcept(false);
+
+    friend class OwnedKeyValueStore;
+};
+
+/**
+ * A database with single writer and multiple readers.
+ * Only the thread that created OwnedKeyValueStore is allowed to invoke `write`.
+ * Creating OwnedKeyValueStore grabs a lock and allows to have consistent view over database.
+ *
+ * To write changes to disk, one must call `commitAndClose` or `disown` with `commitChanges` set to `true`.
+ */
+class OwnedKeyValueStore final {
+    const std::thread::id writerId;
+    std::unique_ptr<KeyValueStore> kvstore;
+    struct TxnState;
+    const std::unique_ptr<TxnState> txnState;
+
+    void clear();
+    void refreshMainTransaction();
+    int commitOrAbortTransactions(bool commit);
+
+public:
+    OwnedKeyValueStore(std::unique_ptr<KeyValueStore> kvstore);
+    ~OwnedKeyValueStore();
+
     /** returns nullptr if not found*/
     u1 *read(std::string_view key);
     std::string_view readString(std::string_view key);
     void writeString(std::string_view key, std::string_view value);
-    /** can only be called from main thread */
+    /** can only be called from owning thread */
     void write(std::string_view key, const std::vector<u1> &value);
-    ~KeyValueStore() noexcept(false);
-    static bool commit(std::unique_ptr<KeyValueStore>);
+
+    /** Convert an owned key value store into an unowned key value store. If `commitChanges` is `true`, writes are
+     * committed to disk. */
+    static std::unique_ptr<KeyValueStore> disown(std::unique_ptr<OwnedKeyValueStore> ownedKvstore, bool commitChanges);
```

Comment:
this is only used with `true` as second argument, right?
Can we remove that optionality and support only the current usecase? 

Also, why does it commit at all? Can we decouple the two?

Reply:
>Also, why does it commit at all? Can we decouple the two and e.g enforce that all changes have been commited in disown?

Once committed, the OwnedKeyValueStore can't be used anymore.

I thought it was cleanest to commit when we "consume" the object so that no one accidentally tries to use a committed kvstore.

Reply:
🤔 Ohhh I can just make `commit` / `abort` methods instead that consume the object and return a KVStore.